wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,386 @@
+	function getWeaponType() returns weapontype
+		return weaponType
+	function getState() returns DamageState
+		return state
+
+	/** If the damage is modified by spell bracers ability. */
+	function isSpell() returns bool
+		return attackType == ATTACK_TYPE_NORMAL
+
+	/** Returns true if the damage is going to kill or already have killed target */
+	function isKillingBlow() returns bool
+		return (state == DamageState.FIRED and killingBlow) or (state != DamageState.FIRED and target.getHP() - .406 < amount)
+
+	function getInitial() returns real
+		return initialAmount
+	/** Returns the actual damage that was withdrawn from the unit.
+		Becomes valid only for listeners and after the event. */
+	function getActual() returns real
+		return actualAmount
+	function getAmount() returns real
+		return amount
+
+	protected static Damage previousDamage = null
+	/** Returns the damage instance that was created on last Damage event.
+		For both listeners and modifiers it is previous to callback argument.
+
+		Allows to retrieve the instance right after unit.damageTarget() call,
+		but it is not guaranteed. */
+	static function getPrevious() returns Damage
+		return previousDamage
+
+	// Manual instance creation
+	
+	/** Allows manual creation of Damage instance, which is then dealt
+		through .deal() call, guaranteed passing it to modifiers and listeners.
+
+		Consider also .saveInstance() and .modify() before it's call. */
+	construct(unit _source, unit _target, real _amount, bool _ranged, bool _attack, attacktype _attackType, damagetype _damageType, weapontype _weaponType, int _data)
+		source = _source
+		target = _target
+		initialAmount = _amount
+		amount = _amount
+		ranged = _ranged
+		attack = _attack
+		attackType = _attackType
+		damageType = _damageType
+		weaponType = _weaponType
+		data = _data
+			
+	protected static Damage dealingInstance = null
+	/** applies user-created Damage instance, allowing it to be retrieved
+		through Damage.retrieve() funciton. */
+	function deal()
+		if state == DamageState.CREATED
+			state = DamageState.FIRING
+			saveInstance() // will be released after processing
+			let temp = dealingInstance
+			dealingInstance = this
+			Log.trace("deal d" + I2S(this castTo int))
+			UnitDamageTarget(source, target, amount, attack, ranged, attackType, damageType, weaponType)
+			Log.trace("completed d" + I2S(this castTo int))
+			dealingInstance = temp
+			if state != DamageState.FIRED
+				Log.debug("Damage.deal() returned not fully processed damage! (state: " + state.toString() + ")")
+				releaseInstance()
+		else
+			Log.debug("Damage.deal() called on already dealt damage!")
+
+	/** Damage instances are automatically destroyed.
+		Use this to protect them if you need an instance anyhow.
+		
+		You need then to use `releaseInstance()`
+		in order to destroy the instance.
+
+		This means that the instance will be destroyed
+		when the last `releaseInstance()` is called on it */
+	function saveInstance()
```

Comment:
Why copy if you can just save it?

Reply:
I made the exact same suggestion, maybe take a hint Trokkin?
This save/release stuff is unintuitive, 
> "Use this to protect them if you need an instance anyhow."

Doesn't really explain why I need this.

Reply:
Main reason this exists is because damage has a lot of different data inside of it and making a storage for it and then copy it from event would be a lot of a extraneous code.
And some additional overhead, if anybody cares about that.

Basically these two functions mimic shared ptr behavior. Maybe if I will redo documentation based on this fact it could become more intuitive?

Reply:
To answer your first question, you copy it to have the guarantee that the original object was cleaned up by the library. Furthermore, removing this will simplify the interface exposed to the user.

As a counter point to your reasoning, the damage object does have a lot of data in it and much of it may not want to be stored by the user of this library. It would be beneficial to them to just grab the data they want and store it themselves.

As another thought on this, the save/release methods of this class don't directly relate to the function of this library, and thus shouldn't be exposed to the user.